optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/Joiners.java

```diff
@@ -50,6 +57,19 @@
         return new SingleBiJoiner<>(leftMapping, JoinerType.EQUAL, rightMapping);
     }
 
+    public static <A> BiJoiner<A, A> notEqual() {
+        return notEqual(Function.identity());
+    }
+
+    public static <A, Property_> BiJoiner<A, A> notEqual(Function<A, Property_> mapping) {
+        return notEqual(mapping, mapping);
+    }
+
+    public static <A, B, Property_> BiJoiner<A, B> notEqual(Function<A, Property_> leftMapping,
+            Function <B, Property_> rightMapping) {
+        return new SingleBiJoiner<>(leftMapping, JoinerType.NOT_EQUAL, rightMapping);
```

Comment:
Will notEqual be indexed?
If it's not indexed, there must be fail-fast if any indexed joiner is behind it (so only notEquals and filtering() can come after it).

Reply:
All in all, the notEquals raises a red flag of "when in doubt, leave it out" for me, especially due to that last question, until we discuss it properly and/or know more about how users use ConstraintStreams. Wdyt?

Reply:
In Drools, NOT_EQUAL is an implemented index type, so CS-D behaves the same for EQUAL as for NOT_EQUAL.

There is no Bavet support for NOT_EQUAL and it will throw an exception if you try to use it.

Reply:
There is no way to leave this out, as otherwise I can't have `ifExistsOther()`. That requires `notEqual()`. (Alternatively, a filter with poor performance.)